### ISSUE TITLE
Installed site IDs not saved if no per-site install routine on multisite

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#703]
https://github.com/WordPoints/wordpoints/issues/404#issuecomment-314212848:

>I've just realized that when there is no per-site install routine to run, the IDs of the sites aren't added to the list of sites on which the entity is installed. This is true both in the new and old code.
>
>Fortunately, this only affects the ranks component in core, since that is the only thing that doesn't have a per-site install routine. But we ought to fill in the site IDs on update.

**Comment by JDGrimes** [comment]
When there is no per-site install routine to run, the there is actually nothing "installed" on each site. Thus, there is nothing to remove on uninstall (the only place the installed site IDs are used), and therefore no need to keep track of these sites. The behavior is actually correct, these aren't sites where the db is installed, only the code is installed there (activated at one time, that is), which is irrelevant.

Of course, as noted in https://github.com/WordPoints/wordpoints/issues/404#issuecomment-321676673:

>Of course, when such an extension did add an installable class [or in this case a per-site install/update/uninstall routine], it should also set the installed site ID as part of that initial update, using the updater that core offers for that purpose. This will add the IDs of the sites to the list, where this update that installed the extension actually ran, and thus where there was actually something to uninstall.

So in short, this doesn't need to be based on whether we have a per-site install routine, but whether we have a per-site *uninstall* routine.

We should probably check to make sure the docs make this clear though.

**Comment by JDGrimes** [comment]
> So in short, this doesn't need to be based on whether we have a per-site install routine, but whether we have a per-site uninstall routine.

That said, the ranks component *does* have per-site uninstall routines (to delete options), so it ought to be saving the list of sites where it is installed.

**Comment by JDGrimes** [comment]
As part of the API our new installer class automatically adds the installed site IDs when the installable has any per-site install routines. However, in this case, it doesn't, so that doesn't help.

There are several options for how we could work around this:

1. Just add the installed site ID inside the install routines getter function. We'd rather not do this though, because these methods aren't supposed to run the install code themselves.
2. Return the routine to add the installed site IDs. This makes sense, but we'd end up triggering this twice in that case: once by the routine we return, and again when the installer adds the site itself.
3. Return a no-op routine that doesn't do anything, just to trigger the installer code to run. Seems a little silly, but prevents the site ID from being added to the list twice.

The third option seems best here.

**Comment by JDGrimes** [comment]
Actually, it turns out that in the new installer we do call `run_per_site()` when the installable is being per-site installed, rather than network installed. This is why our tests show that the ranks component does in fact uninstall itself properly. And the same goes for the old un/installer.

**Comment by JDGrimes** [comment]
That does still leave network installing though, but I guess that would likely be fixed as part of #704 anyway.

**Comment by JDGrimes** [comment]
So where we stand here is that things currently work correctly as far as this goes when we are network active, even though the ticket title is true, and the installed site IDs aren't saved in that case (but see #704, as I said above). That's because the `Intallable::get_installed_site_ids()` method currently does this:

```php
		if ( $this->is_network_installed() ) {
			$site_ids = $this->get_all_site_ids();
		} 
```

And that is the method that the uninstaller calls to get the list of site IDs to uninstall on.

In other words, currently when network active it isn't actually necessary to save the list of site IDs, because all site IDs are run over anyway. This was also true in the old code.

As noted in https://github.com/WordPoints/wordpoints/issues/704#issuecomment-318859561 though:

>However, this behavior assumes that we will always be installed on every single site on multisite when network installed, which is not true. If we have been inactive a while before we are uninstalled, then we won't have been installed on every site, if new ones were created during that time. See https://github.com/WordPoints/wordpoints/issues/404#issuecomment-318191246.

In that referenced comment, we decided based on this to only check count of the list of installed site IDs when deciding whether to skip uninstall.

The next comment in https://github.com/WordPoints/wordpoints/issues/404#issuecomment-318859920 also noted that

>it looks like we don't actually save the site IDs when network active at all. But that will likely change.

So I guess we just never fully completed that, perhaps partly leaving it to these other tickets. (Well, here we are!) In particular, we figured that we wanted to save all of the site IDs at once in that case, as #704 proposes, and so we'd fix finish making that change as part of that.

**Comment by JDGrimes** [comment]
So for clarity:

This behavior is currently entirely correct, the site IDs are being saved as they need to be in order for the code as it is to uninstall properly.

However, we did anticipate changing the way the code works, so that it also relies on this list for network installed entities, whereas it currently does not. In that case, if we do not save the list of all sites IDs as proposed in #704, we won't end up saving them at all in absence of a per-site install routine, and thus uninstall will not run.

That would actually also be entirely correct behavior, if the absence of per-site install routines also indicates an absence of per-site *uninstall* routines—because in that case, no per-site uninstall routines need to be run, so we don't need to know what sites to run them on.

However, for the ranks component, we *do* have per-site uninstall routines, without any per-site install routines, with the result that do need to save the list of all site IDs on install in this case. #704 would cause us to do that.

It should be noted however, that in #704 we don't actually need to save the list of all site IDs if there are no per-site install routines; the correct behavior would actually be to check if there were any per-site install routines before doing this, since otherwise we don't know that anything will actually be installed that we'll need a list to uninstall on later. As such, we'd probably still need special handling for the ranks component.

**Comment by JDGrimes** [comment]
For network active installables, we'd also need to maintain the list by adding the site ID when installing on a newly added site. We didn't do that in the old un/installer, of course, because we didn't use the list at all then. But we do end up adding the site ID in that case in the new code. 

However, we do so regardless of whether there are any per-site install routines. That is potentially broader than necessary, saving site IDs that we might not actually being installing anything on. So perhaps we need to correct that as well.

**Comment by JDGrimes** [comment]
Alternatively, perhaps it would be best to just go ahead and save and maintain the list of installed site IDs anyway, just because there is an installable class for the entity, regardless of whether or not it actually has anything to uninstall for each site at present. It would just be easier to handle this internally rather than requiring each installable to properly manage it by specifying the correct install/update routines to insure this information is saved.

This is a different story from when an extension doesn't offer an installable object at all, which was previously discussed and the idea rejected (https://github.com/WordPoints/wordpoints/issues/404#issuecomment-321676673).

Of course, as noted somewhere, we could actually grab the list of uninstall routines and check if there were any per-site ones, and save the installed site IDs accordingly. The main thing that is holding me back from that is the thought that the uninstall routine getter might be running some uninstall code directly. Which it shouldn't be doing, but we actually do ourselves in the uninstaller for core (we could probably work around or change this though). Also, the factories for the uninstall routines could inadvertently run some code (though they shouldn't either), ~~and they can be fairly expensive, potentially (DB queries for wildcard options).~~ Actually, they don't run the query within the factory, but within the routine itself, as it should be (and must be for per-site stuff anyway). So though they do do a bit of work, it shouldn't be a problem since we'd just be checking it during install and possibly update anyway.

For updates though, would we have a chance to do this when the uninstall routine code was updated? I guess we would, since the updater is triggered by the new installables app for code updates, regardless of whether there are actually any DB updates (it doesn't check that, it is left to the updater to do its thing).

And also on update, how would we know that we needed to add the site IDs though? That is, how would we know that they weren't already added? Well, I guess that for network-active entities that would be obvious, because no site IDs would be present. But for per-site active entities, I guess we'd have to check the list. We could make the list keyed by site ID as well to make this easier, I suppose. 

OK, I can't really think of any more excuses. I guess I'm getting used to the idea. Maybe even beginning to like it.

**Comment by JDGrimes** [comment]
I guess though, that we really ought to pursue that in a separate ticket.

The behavior that this ticket was calling a bug is actually correct, and though we've found some related things that need to be addressed, it doesn't change the fact that this ticket's premise is invalid. So I guess I'll close it as such. The real bug is not that site IDs aren't be saved for installables without per-site install routines, but that *if* we change the behavior, it won't be saved for those that do have per-site *uninstall* routines. As the problem was not identified correctly, and the code that the "bug" was in has actually not yet been written, this ticket is invalid, and merely informative. It was created out of the mistaken thinking that the installed site IDs should represent the list of sites where the code was active, rather than where things were actually installed.
